### ISSUE TITLE
investigate use of llapi_file_get_stripe for alignment on Lustre [REPLACEMENT ISSUE]

### ISSUE THREAD

**Issue by shanedsnyder** [darshan-hpc/darshan#134]
The original issue

	Id: 134
	Title: investigate use of llapi_file_get_stripe for alignment on Lustre

could not be created.
This is a dummy issue, replacing the original one. It contains everything but the original issue description. In case the gitlab repository is still existing, visit the following link to show the original issue:

TODO

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Feb 19, 2016, 17:58

Will look into this as we investigate a Lustre module in Darshan.

Glenn mentioned using /proc/fs/lustre/llite to grab some of the Lustre data we want, so that broadly leaves 2 questions to be answered:

* how does using the llapi functions compare to poking /proc in terms of overhead? 
* do either of the options give access to more than just stripe data that might be useful?

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on May 20, 2016, 13:54

Glenn has been working on this in the lustre-mod branch of Darshan.

An initial cut at the module is complete, but we are waiting on resolution of Issue #187 to close this out. Let's aim to include this in 3.0.2 release, if possible.

**Comment by shanedsnyder** [comment]
In GitLab by @glennklockwood on May 24, 2016, 15:03

I've been thinking about how to implement this and the following approach is the frontrunner:

First, add an `int64_t *` to each `darshan_lustre_record` that points out to that record's array of OST ids, e.g.,

```
struct darshan_lustre_record
{
    darshan_record_id rec_id;
    int64_t rank;
    int64_t counters[LUSTRE_NUM_INDICES];
    int64_t *ost_ids;
};
```

Then, in the `lustre_get_output_data` function, allocate and populate a new buffer that contains `darshan_lustre_record`s with their `ost_ids` pointers replaced with the actual array of OST IDs.  The OST count (and therefore the array length) is already stored in `darshan_lustre_record.counters`, so this serialized representation should then be digestible to the parser.

However there are two problems here:

1. Record data is now being stored outside of the memory buffer allocated in `lustre_runtime_initialize` after `darshan_core_register_module` is called.  Maybe this isn't terrible, seeing as how the runtime record structs (e.g., `posix_file_runtime`) are also living outside of the memory buffer that's bounded by `mem_limit`

2. `get_output_data` on rank 0 now has to allocate a second buffer of size `mem_limit` in which it can build the serialized record representation.  This nominally doubles the memory footprint of this module.

I wasn't sure how careful I should be about staying within the `mem_limit`-sized record buffer.  It seems like `mem_limit` is supposed to limit how much heap a module uses, but at the same time, modules don't seem shy about malloc'ing buffers outside of their official `file_record_array` space.

Would the above approach be terrible since it requires a bunch of heap that is "off the grid" with respect to darshan-core, or should I plan on figuring out a way to keep both the counters and the `ost_ids` arrays in the same memory buffer?

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on May 24, 2016, 16:53

These are good questions but a little complicated to answer right now... the lustre-mod branch is a descendant of master, which handles memory allocation a good deal differently than how it is in the mmap-dev branch (which will become master in the very near future, shooting for this week). In hindsight, maybe it would have made more sense to base lustre-mod off of the mmap branch, but I was worried that would be problematic, too, as I'm still actively developing the core code there.

At any rate, I can try to give you some details that would be helpful that apply to Darshan in general, then give you an idea of how to somewhat future-proof your module for being merged in with the mmap branch. I can also give details on the memory management differences currently between the 2 branches.

So, the master branch just returns a maximum memory limit (`mem_limit`) that should be used by a module, but it has no way of enforcing that. In general, we just try to ensure that the actual records the module is tracking fit entirely in this memory -- that is why most of the modules just divide this memory size by the module record size to determine the maximum number of records the module can track. We don't count the extra mallocs we use for data structures needed for managing these records at runtime though (e.g., runtime records, hash table references, etc.). In a sense, that memory limit bounds the amount of record data per module per process that makes it to the log file, not necessarily the runtime footprint.

In the mmap branch, darshan-core actually mallocs (well, mmaps the buffer if mmap feature is enabled) the big contiguous buffer for that module and manages it internally. When a module registers a record, it must indicate how large it is, so that darshan-core can determine if there is space for it or not. If space is available, a pointer is returned indicating where the record can be written. You can still malloc some helper structures to use at runtime, we don't care as long as its within reason.

Now, on to why the simpler solution you gave above won't work correctly with the mmap branch: when we are mmap'ing the records to a temporary log file, we need to get them persisted in their final format. Since the app might forget to call finalize (or crash), you're not guaranteed you'll get the opportunity in `get_output_data` to correct everything like that. So, in the master branch this strategy would work fine, but I'd recommend getting the entire record stored in a contiguous buffer to set yourself up for an easier merge into the mmap branch. I think what you would have to do is calculate exactly how big your variable-length record will be before registering it, then serializing it in its final format if darshan-core gives you a buffer to use.

Hopefully that makes sense? That got pretty long-winded and probably could be organized better. Let me know if I can clarify anything.

**Comment by shanedsnyder** [comment]
In GitLab by @glennklockwood on May 24, 2016, 18:16

That makes perfect sense, and I like the sound of having darshan-core do the memory allocation for modules.  Seems more rigorous.

Requiring that `darshan_lustre_record` contains a variable-length field leads to another question: how do you prefer modules do this in their record struct definition?

1. using `int64_t array[]`, which is only valid in C99
2. using `int64_t array[0]`, which is a GCC-ism
3. using `int64_t array[1]`, which I guess is the most portable way

(stolen from https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html)

I wasn't sure if Darshan already depends on a C99-compliant compiler, and if it doesn't, I didn't want to rock the boat without asking first.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on May 25, 2016, 10:13

To be honest, I didn't even realize you could do zero-length arrays like that, so I'm not sure I know the answer. But, Darshan doesn't depend on a C99 compiler, so we should probably avoid that. I suppose we also can't depend on something GCC-specific either to support IBM, Intel, etc. compilers. So, I'd suggest 3 based on that.

I've handled a similar problem in darshan-core for storing the variable-length filenames, but was just serializing this data directly into a buffer myself and hanging onto the starting address of the buffer to deserialize later if I needed it. I'll probably mess around with applying option 3 above to see how it works, since that would simplify some existing code.

**Comment by shanedsnyder** [comment]
In GitLab by @glennklockwood on Jun 7, 2016, 17:52

The lustre-mod branch is now able to gather and store OST mappings for each file.  In working on the parsing side, the following assumption:

```
#define DEF_MOD_BUF_SIZE 1024 /* 1 KiB is enough for all current mod records ... */
```

in `darshan-parser.c` might not hold true.  Since each OST mapping is stored as a 64-bit int, this `DEF_MOD_BUF_SIZE` can only hold 128 OSTs before the record overflows.

Should we either

1. increase `DEF_MOD_BUF_SIZE` to be substantially larger--but it'd probably have to be a lot larger, since Lustre supports > 20k OSTs
2. decrease the precision of the OST IDs to 16- or 32-bit ints

or should we do both?  At what OST count should we accept that Darshan's Lustre module won't work?

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Jun 8, 2016, 10:29

That `DEF_MOD_BUF_SIZE` define was just a simple shortcut for getting some of the utility codes up and running, given that I knew all of the record sizes were smaller than that. Basically just a simple way to get records from all of the modules without having logic for determining which has the largest record size, etc. I suspected we could just bump it up to a larger size if modules with bigger records were around, but maybe that's not the right answer for variable-length records. The issue being that the caller of the logutils API may not know the maximum size of a record it might extract (particularly for variable-length records).

Probably the thing to do is just have the logutils API allocate a buffer for storing the record itself, then return it to the user to hold onto rather than requiring the caller pass a big enough buffer with no context on how big it might need to be. I'll open a ticket for doing that, but you can probably just bump that define value up to whatever for your prototyping purposes. 

As far as the OST IDs go, I would just make there datatype size just big enough to hold every possible OST ID just to be safe. A 16-bit integer gives you 32k distinct OST IDs, which seems like more than enough for the Lustre max (and way more than enough for what you'd see in practice, I suppose). You could further make the OST ids unsigned (maybe you are already doing this), and that would buy you 64k instead. We have a need to set some existing Darshan counters to `-1` from time to time (for unsupported counters in different Darshan versions), which is why we typically use signed ints, but that seems unnecessary for storing these IDs.

**Comment by shanedsnyder** [comment]
In GitLab by @glennklockwood on Jun 15, 2016, 16:44

This is now feature complete as of b51c463b59485b4d3f2abd90f63285ff8d72d06f.  The Lustre module now retains the OST IDs for each file in the log, correct reduces shared files, and the parser displays all of the OST IDs correctly.  I did a couple of simple tests under normal conditions, but haven't subjected this code to any weird edge cases.  There are three things that I'm not 100% pleased about:

1. darshan-counter prints one counter for each OST_ID (e.g., `LUSTRE_OST_ID_0`, `LUSTRE_OST_ID_1`, `LUSTRE_OST_ID_2` ...), so the output is very verbose for widely striped files
2. I dialed up `DEF_MOD_BUF_SIZE` to 640 KiB
  - this is enough for 10K OST IDs stored as `int64_t`; we can safely change this to `int32_t` before release though
  - I moved the arrays of size `DEF_MOD_BUF_SIZE` onto the heap for darshan-parser and darshan-convert, but _not_ darshan-diff.  darshan-diff is substantially more complicated than darshan-{diff,convert}, and I reason that most of the potential for users accidentally overflowing their stacks is gone by addressing this in diff/convert.  This does mean that diff now looks less like its brother and sister though.
3. I created a new module-specific runtime header in `darshan-runtime/darshan-lustre.h` for the benefit of a test framework I've started building that stubs out the darshan-core functionality to allow fs-specific modules to be tested outside of Darshan (`darshan-test/lustre/darshan-core-stub.c`).  I'm not sure that this is the best approach, but I didn't want to have to update a bunch of `extern`s in my test framework every time I changed the runtime structures in darshan-runtime.

I'll try to cook up a matrix of edge cases to test to ensure that this module is solid.  Help in testing would be appreciated :smiling_imp:

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Jun 15, 2016, 17:18

All sounds good to me.

We'll have to think some about 1.) eventually. Darshan's output always tends to look a bit verbose to make it easily parsable by other tools/scripts, but I'm not sure of the best way to format the OST data to be easily readable by humans and machines alike. One option is to come up with some compressed way of presenting this data that can be easily reversed to build the OST list (e.g., OST IDs 0, 1, 2, 3, 4, 5, 6, & 7 could be displayed as OST IDs 0--7). 

2.) I can sort out when we merge you back into master branch. I have an open ticket to allocate the records dynamically for the logutils caller which would avoid this problem all together, but I haven't wrote the code yet.

3.) sounds fine for now, but I'd have to look over the code in more detail to understand I think. We certainly won't judge you for being vigilant in your testing :smirk: 

FYI, I'm almost ready to merge mmap-dev back onto master, just working on maintaining backwards compatibility between versions. After that, we can start working on getting this merged over if you think it's ready.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Aug 23, 2016, 16:02

I'm circling back to mark this issue as resolved.

The original issue was to determine how feasible it is to use the `llapi_file_get_stripe`call in the Lustre module for determining the stripe info of each accessed file. Glenn's initial tests indicated that the overhead of the `llapi_file_get_stripe` is relatively high and scales with the job size, precluding us from using it within Darshan wrappers. We were able to confirm that using a Lustre-specific ioctl, we can grab the same data in a low-overhead manner. This functionality is completely merged into the master branch of Darshan and will be included in release 3.1.0.

I have created a new ticket (#200) to consider using new Lustre client side APIs to grab Lustre data rather than using low-level ioctls. This will be a more portable, cleaner way of grabbing this data, but we don't have a system with a new enough Lustre version to verify the overhead of this API. We will continue to track that issue and update in the future.
